File: C++/sort-integers-by-the-number-of-1-bits.cpp
class Solution {
public:
    vector<int> sortByBits(vector<int>& arr) {
        sort(arr.begin(), arr.end(), [this](int a, int b) {
            int na = countBits(a), nb = countBits(b);
            if (na == nb) {
                return a < b;
            } else {
                return na < nb;
            }
        });
        return arr;
    }

    int countBits(int x) {
        int c = 0;
        while (x) {
            c += x & 1;
            x >>= 1;
        }
        return c;
    };

};

File: JavaScript/sort-integers-by-the-number-of-1-bits.js
/**
 * @param {number[]} arr
 * @return {number[]}
 */
var sortByBits = function (arr) {
    arr.sort((a, b) => {
        let ca = 0, cb = 0
        for (let x = a; x > 0; x >>= 1) {
            ca += x & 1
        }
        for (let x = b; x > 0; x >>= 1) {
            cb += x & 1
        }
        if (ca == cb) {
            return a - b
        } else {
            return ca - cb
        }
    })
    return arr
};